### DISCUSSION TITLE
Access to data on both sides of a refinement interface

### DISCUSSION THREAD

**Discussion by avergnaud002** [AMReX-Codes/amrex#3343] [Q&A]
Hello,

Given a cell or a face (with index i,j,k) at a refinement interface, is there a simple way to access the data (cell-centered and/or face-centered data) that are on the other side of the interface ? I am interested of access both fine data from coarse level, and coarse data from fine level.

Potentially those data are on different ranks but I want to avoid, as much as possible, blocking calls and loops over MultiFab and/or boxes.

I want to call this sort of « function » inside a loop (like mfiter or octree_iter), so it is maybe more clever to work directly at the box/Fab « level » (instead of working from i,j,k index).

I use the Fortran Interface of AMReX.

thank you
Alban

**Comment by WeiqunZhang** [comment]
For a given coarse cell next to the coarse/fine interface, there are multiple fine cells on the other side of the interface. Are you trying to access all the fine cells? For a given fine cell next to the interface, do you simply want the value of the coarse cell next to it without any spatial interpolation? Have you looked at the functions in AMReX_fillpatch_mod.F90? If you type `git grep amrex_fillpatch", you should see some examples of using the fillpatch function.

**Reply by avergnaud002** [reply]
Yes, for a given coarse cell I want to access all the fine cells. And yes, for a given fine cell I want the value of the coarse cell without any interpolation.
I will look at the functions in AMReX_fillpatch_mod.F90. I had not thought that we could use these functions to solve my problem ;).
Thank you for your answer.

**Reply by avergnaud002** [reply]
In the use of functions in AMReX_fillpatch_mod.F90, which interpolater corresponds to "no interpolation" ?

**Reply by avergnaud002** [reply]
I also thought of a solution using "parallel_copy".
Given the i,j,k index of the given cell and the refinement ratio, it is easy to define a "small" MultiFab containing only the cell(s) we want to access. Then I want to call parallel_copy to copy date in this MultiFab. Does this solution seems efficient ?

**Reply by WeiqunZhang** [reply]
You can use `amrex_interp_pc` as interpolater. (Here `pc` mean piecewise constant.) That should fill fill multifab's ghost cells outside the coarse/fine interface with coarse data. But fillpatch cannot provide you fine data for the coarse cells in the way you want. Maybe you can let the processes with the fine cells do the work and then copy them to the processes with the coarse cells?

**Reply by avergnaud002** [reply]
Ok for the amrex_interp_pc to access coarse data from fine cell.
Yes I think about this kind of solution for access fine data from coarse cell. Does the solution mentioned above with "ParallelCopy" seems good to you ?

**Reply by WeiqunZhang** [reply]
Sure, you can do that. But you might have to write some C++ code to do it.

**Reply by avergnaud002** [reply]
Ok thank you very much for your answer !
I think "parallel_copy" is available with the Fortran Interface... Or am I missing something ?
